---
title: "Rmarkdown_Week_13_Lecture_1"
author: "Enrique I. Velazquez Villarreal, MD, PhD, MPH, MS"
date: "4/15/2019"
output: html_document
---

## Uploading the file "trgn599.clinical.tsv":

```{R}
#data.dir <- '/Users/enriquevelazquez/Documents/R_working_directory/E-GEOD-31301'

setwd('/Users/enriquevelazquez/Documents/R_working_directory/E-GEOD-31301')

my.dat<-read.delim("GSM775491_sample_table.txt",row.names=1)

adf <- read.delim('A-GEOD-13972.adf.txt',skip=15,row.names=1)
names(adf) <- c("Coordinate","Sequence")

```

```{R}

head(my.dat)

```

```{R}

head(adf)

adf$chr <- apply(matrix(adf$Coordinate,ncol=1),1,function(coord){strsplit(coord,':')[[1]][1]})
adf$pos <- apply(matrix(adf$Coordinate,ncol=1),1,function(coord){strsplit(coord,':')[[1]][2]})
adf$start <- apply(matrix(adf$pos,ncol=1),1,function(coord){strsplit(coord,'-')[[1]][1]})
adf$end <- apply(matrix(adf$pos,ncol=1),1,function(coord){strsplit(coord,'-')[[1]][2]})
adf <- adf[!is.na(adf$chr),]

adf$start <- as.numeric(adf$start)
adf$end <- as.numeric(adf$end)
adf$chr <- as.factor(adf$chr)

```

```{R}
my.dat <- cbind(my.dat,adf[,c(3,5,6)])
head(my.dat)

```

```{R}

table(my.dat$end-my.dat$start+1)

```

```{R}

table(my.dat$chr)

```

```{R}

plot(density(my.dat$VALUE,na.rm=T))

```

```{R}
my.dat.c4 <- my.dat[my.dat$chr=='chr4',]
my.dat.c4 <- my.dat.c4[order(my.dat.c4$start),]

w.start <- min(my.dat.c4$start)
w.end <- max(my.dat.c4$end)

plot(my.dat.c4$VALUE ~ my.dat.c4$start,t='l',xlim=c(w.start,w.end),main="ChIP-Chip peaks\non Chromosome 4",xlab="Position",ylab="log2 IP/Input")
abline(h=0,col="blue",lty=3)

```

```{R}

w.start <- 500000
w.end <- 512000

plot(my.dat.c4$VALUE ~ my.dat.c4$start,t='l',xlim=c(w.start,w.end),main="ChIP-Chip peaks\non Chromosome 4",xlab="Position",ylab="log2 IP/Input")
abline(h=0,col="blue",lty=3)

```

```{R}
w.size <- 1000

w.pos <- rep(NA,(w.end-w.start)/w.size)
w.smooth <- w.pos

i <- 1
  
for(pos in seq(w.start,w.end+w.size,w.size)){
  vals <- my.dat.c4$VALUE[my.dat.c4$start>pos &  my.dat.c4$start<(pos+w.size)]
  w.pos[i] <- pos
  w.smooth[i] <- mean(vals)
  i <- i+1
}

plot(my.dat.c4$VALUE ~ my.dat.c4$start,t='l',xlim=c(w.start,w.end),main="ChIP-Chip peaks\non Chromosome 4",xlab="Position",ylab="log2 IP/Input")
abline(h=0,col="blue",lty=3)
lines(w.pos,w.smooth,col="red",lty=2)
```

```{R}
## replicates

##reporters <- row.names(my.dat.c4)

##my.dat.2<-read.delim("GSM775492_sample_table.txt",row.names=1)
##my.dat.3<-read.delim("GSM775493_sample_table.txt",row.names=1)

##my.dat.c4$rep2 <- my.dat.2[reporters,]
##my.dat.c4$rep3 <- my.dat.3[reporters,]

##plot(my.dat.c4$VALUE,my.dat.c4$rep2)
#plot(my.dat.c4$VALUE,my.dat.c4$rep2,my.dat.c4$rep3)

```

```{R}

##pairs(my.dat.c4[,c(1,5,6)])

```

```{R}

panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor, ...){
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r <- abs(cor(x, y,use="pairwise.complete.obs"))
  txt <- format(c(r, 0.123456789), digits = digits)[1]
  txt <- paste0(prefix, txt)
  if(missing(cex.cor)) cex.cor <- 0.8/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex.cor * r)
}
##pairs(my.dat.c4[,c(1,5,6)],upper.panel=panel.smooth,lower.panel=panel.cor)

plot(my.dat.c4$VALUE ~ my.dat.c4$start,t='l',xlim=c(w.start,w.end),main="ChIP-Chip peaks\non Chromosome 4",xlab="Position",ylab="log2 IP/Input")
abline(h=0,col="blue",lty=3)
lines(my.dat.c4$start,my.dat.c4$rep2,col="red")
lines(my.dat.c4$start,my.dat.c4$rep3,col="green")

```

# High-throughput sequencing of ChIP fragments
```{R}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("EatonEtAlChIPseq", version = "3.8")

library(EatonEtAlChIPseq)

#data.path <- paste0(system.file(package = "EatonEtAlChIPseq"),"/extdata")
data.path <- dir(system.file(package = "EatonEtAlChIPseq"),pattern='extdata',full.names=T)
map.files <- list.files(data.path,pattern=".gz")
map.files

```

```{R}
#/Users/enriquevelazquez/Documents/R_working_directory/
#data.path <- '/home/ortutay/tmp/HTDA/T6_data/chipseq'
#data.path <- '/Users/enriquevelazquez/Documents/R_working_directory/'
library(ShortRead)

chip.aln <- readAligned(data.path, pattern="GSM424494_wt_G2_orc_chip_rep1_S", type="MAQMapview")

chip.aln

```

```{R}

head(chromosome(chip.aln))

```

```{R}
levels(chromosome(chip.aln))
```

```{R}

new.chr <- chromosome(chip.aln)
levels(new.chr) <- "chrXIV"
chip.aln <- renew(chip.aln,chromosome=new.chr)
head(chromosome(chip.aln))

```

```{R}

chip.ranges <- as(chip.aln,"GRanges")
ranges(chip.ranges)
mean(width(ranges(chip.ranges)))
```

```{R}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("chipseq", version = "3.8")

library(chipseq)

estimate.mean.fraglen(chip.ranges,method ="coverage")
estimate.mean.fraglen(chip.ranges,method ="correlation")

chip.ext.ranges <- resize(chip.ranges,width=200)

chip.cov <- coverage(chip.ext.ranges)

plot(as.numeric(dimnames(table(chip.cov))[[2]]),log2(as.numeric(table(chip.cov))),xlim=c(0,3000),xlab="Coverage",ylab="Number of bases (log2)")

```

```{R}

plot(as.numeric(dimnames(table(chip.cov))[[2]]),log2(as.numeric(table(chip.cov))),xlim=c(0,300),xlab="Coverage",ylab="Number of bases (log2)")
```

```{R}
peakCutoff(chip.cov, fdr = 0.003)

chip.peaks <- slice(chip.cov,lower=8)

chip.peaks$chrXIV

```

```{R}
peak.ranges <- peakSummary(chip.peaks)
peak.ranges

```

```{R}
peak.ranges.df <- as.data.frame(peak.ranges)
max(peak.ranges.df$max)

```

```{R}
rougepeak<-which(max(peak.ranges.df$max)==peak.ranges.df$max)
peak.ranges.df <- peak.ranges.df[-rougepeak,]

plot(peak.ranges.df$maxpos,peak.ranges.df$max,type="s",main="ChIP peaks\nChromosome XIV",xlab="Position",ylab="Maximum coverage",col="blue")

```

```{R}
cov.pos <- coverage(chip.ext.ranges[strand(chip.ext.ranges)=='+'])
cov.neg <- coverage(chip.ext.ranges[strand(chip.ext.ranges)=='-'])

peaks.pos <- Views(cov.pos, ranges(chip.peaks))
peaks.neg <- Views(cov.neg, ranges(chip.peaks))

peaks.pos$chrXIV

```

```{R}

coverageplot(peaks.pos$chrXIV[1],peaks.neg$chrXIV[1])

```

```{R}

coverageplot(peaks.pos$chrXIV[5],peaks.neg$chrXIV[5])

## peak.ranges.df <- as.data.frame(peak.ranges)
## peak.ranges.df[which(peak.ranges.df$max>1000),]

```

```{R}

chip.high.peaks <- slice(chip.cov,lower=150)

chip.sel.peaks <- peakSummary(chip.high.peaks$chrXIV[width(peakSummary(chip.high.peaks))>100])

bind.sites <- GRanges(seqnames='chrXIV',ranges=unlist(ranges(chip.sel.peaks)),strand='*')

# reference data from the article

peak.files <- list.files(data.path,pattern=".bed",full.names=T)
ref.peaks <- read.delim(peak.files[1],header=F,col.names=c("chr","start","end","name","p.value"))
ref.peak.ranges <- GRanges(seqnames='chrXIV',ranges=IRanges(start=ref.peaks$start,end=ref.peaks$end),strand='*')

as.data.frame(findOverlaps(bind.sites,ref.peak.ranges)) # great! it is almost identical

```

```{R}

# but which genes are there?

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("TxDb.Scerevisiae.UCSC.sacCer3.sgdGene", version = "3.8")

library(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)
txdb <- TxDb.Scerevisiae.UCSC.sacCer3.sgdGene

seqlevels(txdb)
#seqlevels(txdb, force=TRUE) <- c("chrXIV")

```

```{R}

genelist <- genes(txdb)

head(genelist)

```

```{R}

covers <- findOverlaps(bind.sites,genelist,ignore.strand=T)

head(covers)

```

```{R}

affected.genes <- as.data.frame(genelist[subjectHits(covers)])

head(affected.genes)

```

```{R}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("org.Sc.sgd.db", version = "3.8")

library(org.Sc.sgd.db)

select(org.Sc.sgd.db, as.data.frame(affected.genes)$gene_id, keytype="ORF",columns=c("GENENAME","ENTREZID"))

```

# Analysis of binding site motifs
```{R}
bed.file <- 'GSM1504361_PU.1_ChIPseq_PU.1_B_cells.bed.gz'
peak.tab <- read.delim(bed.file,header=F)
names(peak.tab) <- c('chr','start','end','peakName','score')
dim(peak.tab)

```

```{R}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("rGADEM", version = "3.8")

library(rGADEM)

peak.ranges <- RangedData(IRanges(start=peak.tab$start,end=peak.tab$end),space=paste0('chr',peak.tab$chr))

head(peak.ranges)
```

```{R}

# annotation!
# MGSCv37 is the same as UCSC mm9! find correct bioconductor package
# https://genome.ucsc.edu/FAQ/FAQreleases.html


#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("BSgenome.Mmusculus.UCSC.mm9", version = "3.8")

library(BSgenome.Mmusculus.UCSC.mm9)

#

###gadem <- GADEM(peak.ranges,verbose=1,genome=Mmusculus) # takes long-long time 3 hours on my computer

###gadem

```

```{R}

###nMotifs(gadem)

```

```{R}

###nOccurrences(gadem)

```

```{R}

###consensus(gadem)

```

```{R}

###bestmotif <- which(nOccurrences(gadem)==max(nOccurrences(gadem)))

###consensus(gadem)[bestmotif]

```

```{R}

###getPWM(gadem)
###motifs <- getPWM(gadem)

###seqLogo(motifs[[bestmotif]])

```

```{R}

###seqLogo(reverseComplement(motifs[[bestmotif]]))

```

```{R}

###save(gadem,file="pu1_gadem.rda")
###load("pu1_gadem.rda")

```

```{R}
############ run only on a subset of peaks
## 50 best

peak.tab.sel <- peak.tab[peak.tab$score %in% tail(sort(peak.tab$score),50),]

peak.ranges.sel <- RangedData(IRanges(start=peak.tab.sel$start,end=peak.tab.sel$end),space=paste0('chr',peak.tab.sel$chr))

gadem.sel <- GADEM(peak.ranges.sel,verbose=1,genome=Mmusculus)


```

```{R}

gadem.sel

```


```{R}
nMotifs(gadem.sel)
```

```{R}
nOccurrences(gadem.sel)
```

```{R}
consensus(gadem.sel)
```

```{R}
bestmotif <- which(nOccurrences(gadem.sel)==max(nOccurrences(gadem.sel)))
```

```{R}
consensus(gadem.sel)[bestmotif]
```


```{R}

motifs.sel <- getPWM(gadem.sel)

seqLogo(motifs.sel[[bestmotif]])
```

```{R}
seqLogo(reverseComplement(motifs.sel[[bestmotif]]))

```

