---
title: "Rmarkdown_TRGN599_Week_11_Lecture_1"
author: "Eugene Zhou, MS. Instructor: Enrique I. Velazquez Villarreal, MD, PhD, MPH, MS"
date: "3/25/2019"
output: html_document
---

# Next-Generation Sequencing in Transcriptomics

# RNAseq

# Preparing count tables

```{R}
# install package: http://bioconductor.org/packages/release/data/experiment/html/RNAseqData.HNRNPC.bam.chr14.html

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("RNAseqData.HNRNPC.bam.chr14", version = "3.8")

library(RNAseqData.HNRNPC.bam.chr14)

alnFiles <- RNAseqData.HNRNPC.bam.chr14_BAMFILES

exp.des <- factor(rep(c('Control','HNRNPC knockdown'),each=4))

names(exp.des) <- RNAseqData.HNRNPC.bam.chr14_RUNNAMES
```

# Alignment files to read counts

```{R}
# install package: http://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html

# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("GenomicFeatures", version = "3.8")

# install package: https://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene", version = "3.8")

library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
seqlevels(txdb)
```

```{R}
seqlevels(txdb) <- c("chr14")
genelist <- genes(txdb)
txlist <- transcripts(txdb)
exonlist <- exons(txdb)
```

```{R}
genelist
```

```{R}
txlist
```

```{R}
exonlist
```


```{R}
# Calculate counts per gene

library(GenomicAlignments)

counts <- data.frame(row.names=genelist$gene_id)

for (i in 1:length(RNAseqData.HNRNPC.bam.chr14_BAMFILES)){
  print(paste("Reading:",alnFiles[i]))
  aln <- readGAlignments(alnFiles[i])
  reads <- GRanges(seqnames='chr14',ranges=ranges(aln),strand='*')
  genecounts<-countOverlaps(genelist, reads)
  a <- data.frame(genecounts)
  names(a) <- RNAseqData.HNRNPC.bam.chr14_RUNNAMES[i]
  counts <- cbind(counts,a)
}
```

```{R}
head(counts)
```

```{R}
dim(counts)
```


```{R}
## eliminate genes without any reads

sum(rowSums(counts)==0)
sum(rowSums(counts)<10)

counts <- counts[!rowSums(counts)==0,]
counts1 <- counts

## Add Symbols as row names
```

```{R}
# Install package: https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("org.Hs.eg.db", version = "3.8")

library(org.Hs.eg.db)

sym <- select(org.Hs.eg.db, row.names(counts), keytype="ENTREZID","SYMBOL")
row.names(sym) <- sym$ENTREZID
sym$SYMBOL[is.na(sym$SYMBOL)]<-sym$ENTREZID[is.na(sym$SYMBOL)]
counts <- cbind(sym$SYMBOL,counts)
row.names(counts) <- as.character(counts$"sym$SYMBOL")
counts$"sym$SYMBOL" <- NULL
```


```{R}
counts["HNRNPC",]
```
```{R}
counts["HSPA2",]
```

#t.test(counts["SPTLC2",1:4],counts["SPTLC2",5:8],alternative="less")
```{R}
# Creating a table
write.table(counts,file="HNRNPC_chr14_gene_read_counts.txt")
# data is ready
```

# Differential expression in simple comparison

```{R}
# Differential expression calculation with edgeR
counts <- read.table("HNRNPC_chr14_gene_read_counts.txt")
```

# A naive t-test approach

```{R}
# naive t test approach
counts["SPTLC2",]
```

```{R}
t.test(counts["SPTLC2",1:4],counts["SPTLC2",5:8])
```

```{R}
tt.p.vals <- apply(counts,1,function(dat){t.test(x = dat[1:4], y = dat[5:8])$p.value})
tt.adj.p.vals <- p.adjust(tt.p.vals,method="fdr")
counts <- cbind(counts,tt.p.vals,tt.adj.p.vals)

head(counts[order(tt.adj.p.vals),])
```


```{R}
counts <- counts[order(tt.adj.p.vals),]
gsign.tt <- counts[counts$tt.adj.p.vals<0.03,]

dim(gsign.tt)
```

```{R}
gsign.tt["BDKRB1",]
```

```{R}
counts["SPTLC2",]
```

```{R}
t.test(counts["SPTLC2",1:4],counts["SPTLC2",5:8], alternative = "less")
```

# Single factor analysis with edgeR

```{R}
# Installing packages: https://bioconductor.org/packages/release/bioc/html/edgeR.html

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("edgeR", version = "3.8")

# exact test with edgeR
counts <- read.table("HNRNPC_chr14_gene_read_counts.txt")

library(edgeR)
library(RNAseqData.HNRNPC.bam.chr14)
exp.des <- factor(rep(c('Control','HNRNPC knockdown'),each=4))
names(exp.des) <- RNAseqData.HNRNPC.bam.chr14_RUNNAMES
genexp <- DGEList(counts=counts,group=exp.des)
genexp
```


```{R}
genexp <- calcNormFactors(genexp)
genexp
```

```{R}
genexp <- estimateCommonDisp(genexp)
genexp
```

```{R}
genexp <- estimateTagwiseDisp(genexp)
genexp
```

```{R}
head(genexp$counts[order(genexp$tagwise.dispersion),])
```

```{R}
tail(genexp$counts[order(genexp$tagwise.dispersion),])
```

```{R}
genediff <- exactTest(genexp)
genediff
```

```{R}
topTags(genediff)
```

```{R}
genediff$table <- cbind(genediff$table,FDR=p.adjust(genediff$table$PValue,method="fdr"))
gsign <- genediff$table[genediff$table$FDR<0.03,]
gsign <- gsign[order(gsign$FDR),]
dim(gsign)
```

```{R}
head(gsign)
```

```{R}
head(gsign, 10)
```

```{R}
head(gsign.tt, 10)
```

## Differential expression with DESeq
```{R}
# Installing package: https://bioconductor.org/packages/release/bioc/html/DESeq.html

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("DESeq", version = "3.8")

#Installing package lattice
#install.packages("lattice")

library(DESeq)

count.set <- newCountDataSet(counts,exp.des)
count.set
```

```{R}
count.set <- estimateSizeFactors(count.set)
sizeFactors(count.set)

plot(genexp$samples$norm.factors,sizeFactors(count.set),xlab="edgeR",ylab="DESeq",main="Normalization factors",type="n")
abline(0,1,lty=2)
text(genexp$samples$norm.factors,sizeFactors(count.set),labels=names(sizeFactors(count.set)),cex=0.8)
```

```{R}
# dispersion estimation
count.set <- estimateDispersions(count.set)
par(mfrow=c(1,3))
plotDispEsts(count.set,main="DESeq: genewise dispersion",cex=1)
```

```{R}
plotBCV(genexp,main="edgeR: genewise dispersion normalized counts",cex=1)
par(mfrow=c(1,1))
```



```{R}
# Dispersion values per gene
head(cbind(genexp$tagwise.dispersion,fData(count.set)))
```

```{R}
plot(cbind(genexp$tagwise.dispersion,fData(count.set)),ylim=c(0,2),main="Gene-wise dispersion",xlab="edgeR",ylab="DESeq")
abline(0,1,lty=2)
```

```{R}
# simple - differential expression
genediff.ds <- nbinomTest(count.set,"Control","HNRNPC knockdown")
plotMA(genediff.ds)
```

```{R}
plot(log2FoldChange ~ baseMean,data=genediff.ds[genediff.ds$padj>0.1,], log="x",xlab="Mean of normalized counts", ylab="log fold change", main="MA plot", pch=20)
abline(h=0,lwd=2,col="red")
points(genediff.ds[genediff.ds$padj<0.1, c("baseMean","log2FoldChange")], col="red")
```

```{R}
gsign.ds <- genediff.ds[genediff.ds$padj<0.03,]
gsign.ds <- gsign.ds[order(gsign.ds$padj),]
row.names(gsign.ds)<-gsign.ds$id
```

```{R}
head(gsign.ds,10)
```

```{R}
head(gsign,10)
```

```{R}
par(mfrow=c(1,3))
plot(gsign[intersect(row.names(gsign),gsign.ds$id),"logFC"],gsign.ds[intersect(row.names(gsign),gsign.ds$id),"log2FoldChange"],main="Log Fold Changes",xlab="edgeR",ylab="DESeq")
```

```{R}
plot(gsign[intersect(row.names(gsign),gsign.ds$id),"FDR"],gsign.ds[intersect(row.names(gsign),gsign.ds$id),"padj"],main="Adjusted p-values",xlab="edgeR",ylab="DESeq")
```

```{R}
plot(gsign[intersect(row.names(gsign),gsign.ds$id),"FDR"],gsign.ds[intersect(row.names(gsign),gsign.ds$id),"padj"],main="Adjusted p-values",xlab="edgeR",ylab="DESeq",log="xy",xlim=c(1e-20,1e-1),ylim=c(1e-20,1e-1))
```

```{R}
par(mfrow=c(1,3))
plot(gsign[intersect(row.names(gsign),gsign.ds$id),"logFC"],gsign.ds[intersect(row.names(gsign),gsign.ds$id),"log2FoldChange"],main="Log Fold Changes",xlab="edgeR",ylab="DESeq")
plot(gsign[intersect(row.names(gsign),gsign.ds$id),"FDR"],gsign.ds[intersect(row.names(gsign),gsign.ds$id),"padj"],main="Adjusted p-values",xlab="edgeR",ylab="DESeq")
plot(gsign[intersect(row.names(gsign),gsign.ds$id),"FDR"],gsign.ds[intersect(row.names(gsign),gsign.ds$id),"padj"],main="Adjusted p-values",xlab="edgeR",ylab="DESeq",log="xy",xlim=c(1e-20,1e-1),ylim=c(1e-20,1e-1))
par(mfrow=c(1,1))
```

#######

# Complex experimental arrangements
```{R}
# Experimental factors and design matrix
repls <- rep(rep(1:2,each=2),2)
design <- model.matrix(~exp.des+repls)
design
```

# GLM with edgeR
```{R}
genexp.glm <- DGEList(counts=counts)
genexp.glm <- estimateGLMCommonDisp(genexp.glm,design)
#genexp.glm <- estimateGLMTrendedDisp(genexp.glm,design)
genexp.glm <- estimateGLMTagwiseDisp(genexp.glm,design)
gene.exp.fit <- glmFit(genexp.glm,design)
gene.exp.fit
```

```{R}
genediff.lrt <- glmLRT(gene.exp.fit,coef=2)
topTags(genediff.lrt)
```

```{R}
genediff.lrt.2 <- glmLRT(gene.exp.fit,coef=3)
topTags(genediff.lrt.2)
```

```{R}
genediff.lrt$table <- cbind(genediff.lrt$table,FDR=p.adjust(genediff.lrt$table$PValue,method="fdr"))
gsign.lrt <- genediff.lrt$table[genediff.lrt$table$FDR<0.03,]
gsign.lrt <- gsign.lrt[order(gsign.lrt$FDR),]

dim(gsign.lrt)
```

```{R}
head(gsign.lrt)
```

```{R}
head(gsign.lrt, 10)
```


```{R}
head(gsign)
```

```{R}
head(gsign, 10)
```

# GLM with DESeq
```{R}
# multifactor fifferential expression with glm
exp.des.cmplx<-data.frame(Type=as.character(exp.des),Repl=repls)
row.names(exp.des.cmplx)<-names(exp.des)
exp.des.cmplx
```

```{R}
count.set.glm <- newCountDataSet(counts,exp.des.cmplx)
count.set.glm <- estimateSizeFactors(count.set.glm)
count.set.glm <- estimateDispersions(count.set.glm)
```

```{R}
# fit models to count data
genediff.ds.mod1 <- fitNbinomGLMs(count.set.glm, count ~ Type + Repl)
genediff.ds.mod2 <- fitNbinomGLMs(count.set.glm, count ~ Type)

genediff.ds.glm.pvals<-nbinomGLMTest(genediff.ds.mod1,genediff.ds.mod2)
genediff.ds.glm.fdr<-p.adjust(genediff.ds.glm.pvals,method="fdr")

genediff.ds.glm <- cbind(genediff.ds.mod1[,2:3],PVal=genediff.ds.glm.pvals,FDR=genediff.ds.glm.fdr)

gsign.ds.glm <- genediff.ds.glm[genediff.ds.glm$FDR<0.03,]
gsign.ds.glm <- gsign.ds.glm[order(gsign.ds.glm$FDR),]
```



```{R}
# compare to edgeR

common.genes <- intersect(row.names(gsign.lrt),row.names(gsign.ds.glm))
dim(gsign.lrt)
```

```{R}
dim(gsign.ds.glm)
```

```{R}
plot(gsign.lrt[common.genes,"logFC"],gsign.ds.glm[common.genes,"TypeHNRNPC knockdown"],main="Fold cahnge ith GLMs",xlab="edgeR",ylab="DESeq")
abline(0,1,lty=2)
```

# Heatmap visualization
```{R}
# heatmap visualization
diff.exp <- data.frame(edgeR=gsign[intersect(row.names(gsign),gsign.ds$id),"logFC"],DESeq=gsign.ds[intersect(row.names(gsign),gsign.ds$id),"log2FoldChange"])
row.names(diff.exp) <- intersect(row.names(gsign),gsign.ds$id)

diff.exp[is.infinite(diff.exp$DESeq),]<--2
```

```{R}
library(gplots) 
my.palette <-rev(redgreen(75))

# full dataset with edgeR results
heatmap.2(log2(as.matrix(counts[rownames(gsign),]+1)),col=my.palette,margins=c(10,12))
```

```{R}
# Behaviour of differentially expressed genes
heatmap.2(as.matrix(diff.exp),Colv=F,col=my.palette,margins = c(12, 10))
```

